UserService_EditUser

lalagvana opened the pull request:
Done:
- Added edit user endpoint + tests.

lalagvana committed:
Added edit user endpoint + tests

KostinP commented on DigitalOffice/Services/UserService/RestRequests/EditUserRequest.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace LT.DigitalOffice.UserService.RestRequests
+{
+    public class EditUserRequest
```
If a method for editing a user is needed for an ordinary user, then it is strange to change the mail and password here, because we have planned separate methods for this. Who is this endpoint for?

lalagvana replied:
Not only for ordinary user but for admin also.

KostinP replied:
I think you didn't understand my previous message. I wanted to ask, is this endpoint needed for a regular user? But apparently this method is more likely for a conditional administrator.

KostinP replied:
This method is not suitable for an ordinary user, the password should not be changed at the click of a finger

KostinP replied:
(maybe there is a solution here) [link](https://docs.fluentvalidation.net/en/latest/custom-validators.html)

lalagvana replied:
I also have to check if a user which calls this method has Right #1 _or isAdmin_. Sooo it's for user and admin both. We just can hide "changing" password or mail for user there but unhide it for an admin.

KostinP commented on DigitalOffice/Services/UserService/Validators/EditUserRequestValidator.cs:
```diff
@@ -0,0 +1,57 @@
+{
+    public class EditUserRequestValidator : AbstractValidator<EditUserRequest>
+    {
+        public EditUserRequestValidator()
```
Hmm, we have the same checks when creating a user. Not good (DRY)

lalagvana replied:
Agree. But is it good to use UserCreateRequestValidator in edit user?

KostinP replied:
It is not necessary to use it, you can separate

KostinP replied:
[link (custom validator)](https://docs.fluentvalidation.net/en/latest/start.html#complex-properties)

MihailKulikov commented on DigitalOffice/Tests/UserServiceTests/Commands/EditUserCommandTests.cs:
```diff
@@ -0,0 +1,95 @@
+                IsAdmin = false
+            };
+
+            Assert.Throws<ValidationException>(() => command.Execute(request));
```
It would be cool to also check that the repositoryMock method did not get called

lalagvana replied:
Fixed

MihailKulikov replied:
Not fixed :(

lalagvana replied:
Hmm, strange. The github repository didn't catch my changes :(

lalagvana replied:
Fixed

MihailKulikov commented on DigitalOffice/Tests/UserServiceTests/Commands/EditUserCommandTests.cs:
```diff
@@ -0,0 +1,95 @@
+                IsAdmin = false
+            };
+
+            Assert.Throws<Exception>(() => command.Execute(request), "Email is already taken.");
```
It would be cool to also check that the repositoryMock method did not get called

lalagvana replied:
Fixed

MihailKulikov reviewed (changes requested):
You forgot to bind new interfaces in Startup.cs

lalagvana committed:
Bind new interfaces in Startup.cs

KostinP commented on DigitalOffice/Tests/UserServiceTests/Commands/EditUserCommandTests.cs:
```diff
@@ -0,0 +1,95 @@
+        [Test]
+        public void ShouldThrowExceptionWhenUserDataIsInvalid()
+        {
+            var request = new EditUserRequest
```
Move this object to SetUp.

lalagvana replied:
Fixed

KostinP commented on DigitalOffice/Tests/UserServiceTests/Commands/EditUserCommandTests.cs:
```diff
@@ -0,0 +1,95 @@
+                .Setup(x => x.EditUser(It.IsAny<DbUser>()))
+                .Throws(new Exception("Email is already taken."));
+
+            var request = new EditUserRequest
```
Move this object to SetUp.

lalagvana replied:
Fixed

KostinP commented on DigitalOffice/Tests/UserServiceTests/Commands/EditUserCommandTests.cs:
```diff
@@ -0,0 +1,95 @@
+                .Setup(x => x.EditUser(It.IsAny<DbUser>()))
+                .Returns(true);
+
+            var request = new EditUserRequest
```
Move this object to SetUp.

lalagvana replied:
Fixed

KostinP commented on DigitalOffice/Tests/UserServiceTests/Mappers/EditUserRequestToDbMapperTests.cs:
```diff
@@ -0,0 +1,61 @@
+        [Test]
+        public void ShouldReturnNewDbUserWhenDataCorrect()
+        {
+            var request = new EditUserRequest()
```
Move this object to SetUp.

djino97 reviewed (commented):
Add broker logic for check if user which call this method have Right #1 or isAdmin

lalagvana committed:
Moved tests object to SetUp method

lalagvana committed:
Deleted messages after NotEmpty

KostinP commented on DigitalOffice/Services/UserService/Mappers/EditUserRequestToDbMapper.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace LT.DigitalOffice.UserService.Mappers
+{
+    public class EditUserRequestToDbUserMapper : IMapper<EditUserRequest, DbUser>
```
Why such a name?

lalagvana replied:
Fixed, I've united two mappers in one UserMapper

KostinP commented on DigitalOffice/Services/UserService/Mappers/EditUserRequestToDbMapper.cs:
```diff
@@ -0,0 +1,34 @@
+                MiddleName = request.MiddleName,
+                Status = request.Status,
+                PasswordHash = (new SHA512Managed().ComputeHash(Encoding.Default.GetBytes(request.Password))).ToString(),
+                AvatarFileId = null,
```
Whenever a user is updated, their avatar disappears?

lalagvana replied:
Fixed

KostinP commented on DigitalOffice/Services/UserService/Mappers/EditUserRequestToDbMapper.cs:
```diff
@@ -0,0 +1,34 @@
+                Status = request.Status,
+                PasswordHash = (new SHA512Managed().ComputeHash(Encoding.Default.GetBytes(request.Password))).ToString(),
+                AvatarFileId = null,
+                IsActive = true,
```
Anytime the user is updated, is it back to work?

lalagvana replied:
Fixed

lalagvana committed:
United two mappers in one class

lalagvana committed:
Added some fields to request model and fixed smth connected with it

KostinP commented on DigitalOffice/Services/UserService/Commands/Interfaces/IEditUserCommand.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace LT.DigitalOffice.UserService.Commands.Interfaces
+{
+    public interface IEditUserCommand
```
Add XML.

lalagvana replied:
Fixed

KostinP commented on DigitalOffice/Services/UserService/Commands/EditUserCommand.cs:
```diff
@@ -0,0 +1,34 @@
+        {
+            validator.ValidateAndThrow(request);
+            var user = mapper.Map(request);
+            return repository.EditUser(user);
```
Add empty line.

lalagvana replied:
Fixed

KostinP commented on DigitalOffice/Services/UserService/Controllers/UserController.cs:
```diff
@@ -21,5 +21,12 @@ public bool CreateUser([FromServices] IUserCreateCommand command, [FromBody] Use
         }
+
+        [Route("editUser")]
+        [HttpPost]
```
Can be combined.

lalagvana replied:
Fixed

KostinP commented on DigitalOffice/Tests/UserServiceTests/Validators/EditUserRequestValidatorTests.cs:
```diff
@@ -0,0 +1,113 @@
+        public void ShouldThrowValidationExceptionWhenMiddleNameTooLong()
+        {
+            var middleName = "Example" + new string('a', 30);
+            validator.ShouldHaveValidationErrorFor(x => x.MiddleName, middleName);
```
Add empty line.

lalagvana replied:
Fixed

KostinP commented on DigitalOffice/Tests/UserServiceTests/Validators/EditUserRequestValidatorTests.cs:
```diff
@@ -0,0 +1,113 @@
+        public void ShouldThrowValidationExceptionWhenEmailTooLong()
+        {
+            var email = new string('a', 50) + "@gmail.com";
+            validator.ShouldHaveValidationErrorFor(x => x.Email, email);
```
Add empty line.

lalagvana replied:
Fixed

KostinP commented on DigitalOffice/Tests/UserServiceTests/Validators/EditUserRequestValidatorTests.cs:
```diff
@@ -0,0 +1,113 @@
+        public void ShouldThrowValidationExceptionWhenStatusTooLong()
+        {
+            var status = new string('a', 300) + "@gmail.com";
+            validator.ShouldHaveValidationErrorFor(x => x.Status, status);
```
Add empty line.

lalagvana replied:
Fixed

KostinP commented on DigitalOffice/Tests/UserServiceTests/Validators/EditUserRequestValidatorTests.cs:
```diff
@@ -0,0 +1,113 @@
+        public void ShouldThrowValidationExceptionWhenAllFieldsAreEmpty()
+        {
+            var user = new EditUserRequest();
+            Assert.Throws<ValidationException>(() => validator.ValidateAndThrow(user));
```
Add empty line.

lalagvana replied:
Fixed

KostinP commented on DigitalOffice/Tests/UserServiceTests/Validators/EditUserRequestValidatorTests.cs:
```diff
@@ -0,0 +1,113 @@
+                Password = "Example",
+                IsAdmin = false
+            };
+            Assert.DoesNotThrow(() => validator.ValidateAndThrow(request));
```
Add empty line.

lalagvana replied:
Fixed

lalagvana committed:
Fix

lalagvana committed:
Merge branch 'develop' into feature/UserService_EditUser

lalagvana committed:
Fix again

lalagvana committed:
Added verifying that repositiory didn't get called

lalagvana committed:
Added XML

KostinP commented on DigitalOffice/Services/UserService/Validators/EditUserRequestValidator.cs:
```diff
@@ -0,0 +1,52 @@
+{
+    public class EditUserRequestValidator : AbstractValidator<EditUserRequest>
+    {
+        public EditUserRequestValidator()
```
I think it's wrong to force the frontend to fill in all fields when editing.

lalagvana replied:
Hmm, I thought in our previous discussion we decided not to change it in case of potentially existing another solution.

lalagvana committed:
Small but important fixes

lalagvana committed:
Fixed command tests

lalagvana committed:
Added controller tests

lalagvana committed:
Lots of fixes

- Fixed namespace of model EditUserRequest.
- Deleted extra usings.
- Moved EditUserRequest to DbUser mapper tests to general UserMapperTests.cs.
- Fixed tests, added verification that in DB there wasn't any changes/ were needed changes.

MihailKulikov commented on DigitalOffice/Tests/UserServiceTests/Commands/EditUserCommandTests.cs:
```diff
@@ -0,0 +1,113 @@
+                .Setup(x => x.EditUser(It.IsAny<DbUser>()))
+                .Returns(true);
+
+            Assert.IsTrue(command.Execute(request));
```
Add verifying for repositoryMock

lalagvana replied:
Fixed

MihailKulikov commented on DigitalOffice/Tests/UserServiceTests/Validators/EditUserRequestValidatorTests.cs:
```diff
@@ -0,0 +1,118 @@
+
+namespace LT.DigitalOffice.UserServiceUnitTests.Validators
+{
+    public class EditUserRequestValidatorTests
```
You are checking not all rules. Check CreateUserRequestValidatorTests in refactoring pull request.

lalagvana replied:
Fixed

MihailKulikov commented on DigitalOffice/Services/UserService/Mappers/UserMapper.cs:
```diff
@@ -37,6 +40,28 @@ public User Map(DbUser dbUser)
+                LastName = request.LastName,
+                MiddleName = request.MiddleName,
+                Status = request.Status,
+                PasswordHash = (new SHA512Managed().ComputeHash(Encoding.Default.GetBytes(request.Password))).ToString(),
```
??? It is not working

lalagvana replied:
Fixed

MihailKulikov commented on DigitalOffice/Tests/UserServiceTests/Repositories/UserRepositoryTests.cs:
```diff
@@ -123,5 +131,59 @@ public void ShouldThrowExceptionWhenEmailIsAlreadyTaken()
+                AchievementsIds = new Collection<DbUserAchievement>()
+            };
+
+            Assert.Throws<Exception>(() => repository.EditUser(user), ExceptionMessage);
```
Verify state of db

lalagvana replied:
Fixed

MihailKulikov commented on DigitalOffice/Tests/UserServiceTests/Repositories/UserRepositoryTests.cs:
```diff
@@ -123,5 +131,59 @@ public void ShouldThrowExceptionWhenEmailIsAlreadyTaken()
+            dbContext.Entry(user).State = EntityState.Modified;
+
+            Assert.True(repository.EditUser(user));
+            Assert.That(dbContext.Users.Find(userId).Equals(user));
```
Verify state of db

lalagvana replied:
Fixed

lalagvana committed:
Fixed validator and tests

MihailKulikov commented on DigitalOffice/Services/UserService/Validators/EditUserRequestValidator.cs:
```diff
@@ -0,0 +1,40 @@
+{
+    public class EditUserRequestValidator : AbstractValidator<EditUserRequest>
+    {
+        public EditUserRequestValidator()
```
Add validation rule for Id and add test for it

lalagvana replied:
Fixed

lalagvana committed:
New validation and test for Id

lalagvana committed:
Small fix for validator

lalagvana committed:
Fixed test

djino97 commented on DigitalOffice/Tests/UserServiceTests/Controllers/UserControllerTests.cs:
```diff
@@ -0,0 +1,46 @@
+using Moq;
+using NUnit.Framework;
+using System;
+using System.Collections.Generic;
```
Not used

lalagvana replied:
Fixed

djino97 commented on DigitalOffice/Tests/UserServiceTests/UserServiceUnitTests.csproj:
```diff
@@ -21,6 +21,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\..\Services\CompanyService\CompanyService.csproj" />
```
Not used

lalagvana replied:
Fixed

lalagvana committed:
Small fixes

lalagvana committed:
Merge remote-tracking branch 'origin/develop' into feature/UserService_EditUser

lalagvana committed:
Merge remote-tracking branch 'origin/develop' into feature/UserService_EditUser

KostinP commented on DigitalOffice/Services/UserService/Validators/UserCreateRequestValidator.cs:
```diff
@@ -8,50 +8,45 @@ public class UserCreateRequestValidator : AbstractValidator<UserCreateRequest>
-                .WithMessage("First name is too short.")
-                .Matches("^[A-Z][a-z]+$")
-                .WithMessage("First name with error.");
+                    .NotEmpty()
+                    .MaximumLength(32)
+                    .WithMessage("First name is too long.")
+                    .MinimumLength(2)
+                    .WithMessage("First name is too short.")
+                    .Matches("^[A-Z][a-z]+$")
+                    .WithMessage("First name with error.");
```
Why?

KostinP commented on DigitalOffice/Tests/UserServiceTests/Commands/EditUserCommandTests.cs:
```diff
@@ -0,0 +1,115 @@
+                IsActive = request.IsActive,
+                AvatarFileId = request.AvatarFileId
+            };
+
```
Extra empty line.

KostinP commented on DigitalOffice/Services/UserService/Validators/UserCreateRequestValidator.cs:
```diff
@@ -8,50 +8,45 @@ public class UserCreateRequestValidator : AbstractValidator<UserCreateRequest>
             RuleFor(user => user.MiddleName)
-                .NotEmpty()
-                .WithMessage("Middle name is empty.")
-                .MaximumLength(32)
-                .WithMessage("Middle name is too long.")
-                .MinimumLength(2)
-                .WithMessage("Middle name is too short.")
-                .Matches("^[A-Z][a-z]+$")
-                .WithMessage("Middle name with error.");
+                    .NotEmpty()
+                    .MaximumLength(32)
+                    .WithMessage("Middle name is too long.")
+                    .MinimumLength(2)
+                    .WithMessage("Middle name is too short.")
+                    .Matches("^[A-Z][a-z]+$")
+                    .WithMessage("Middle name with error.");
 
             RuleFor(user => user.Email)
-                .NotEmpty()
-                .WithMessage("Email is empty.")
-                .MaximumLength(50)
-                .WithMessage("Email is too long.")
-                .EmailAddress()
-                .WithMessage("Email is invalid.");
+                    .NotEmpty()
+                    .MaximumLength(50)
+                    .WithMessage("Email is too long.")
+                    .EmailAddress()
+                    .WithMessage("Email is invalid.");
 
             RuleFor(user => user.Status)
-                .MaximumLength(300)
-                .WithMessage("Status is too long.");
+                    .MaximumLength(300)
+                    .WithMessage("Status is too long.");
 
             RuleFor(user => user.Password)
-                .NotEmpty()
-                .WithMessage("Password is empty.");
+                    .NotEmpty();
```
Why? Fix it.
Re-wrote the comment to highlight any incorrect lines

lalagvana committed:
Merge branch 'develop' into feature/UserService_EditUser

lalagvana committed:
Fixes after merge with develop

lalagvana committed:
Merge branch 'develop' into feature/UserService_EditUser

lalagvana committed:
Fixes after merge with develop

lalagvana committed:
Small fix

lalagvana committed:
Deleted controller tests

lalagvana committed:
Merge branch 'develop' into feature/UserService_EditUser

lalagvana committed:
Fixes after merge with develop

lalagvana committed:
Merge branch 'develop' into feature/UserService_EditUser
